samples/matter/common/src/watchdog/watchdog.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#pragma once
+
+#include <cstdint>
+
+#include "util/finite_map.h"
+
+#include <zephyr/drivers/watchdog.h>
+#include <zephyr/kernel.h>
+
+namespace Nrf
+{
+/**
+ * @brief Watchdog class dedicated to detect code blocking and infinite loops.
+ *
+ * while the new Watchdog object is creating the global Watchdog instance creates a new channel.
+ * Each watchdog object contains own window ranges so all objects may work separately and be fed in a
+ * application-specific way.
+ *
+ * You can feed a Watchdog object in a two ways:
+ * 	- Manually, by calling the Feed() method in a specific context.
+ *	- Automatically, by providing callback function and feed interval value to the constructor as an argument.
+ * 	  If the callback is provided it will be called in each feed interval time periodically.
+ *
+ * @warning All Watchdog object must be created before calling Enable static method.
+ *
+ * To start the global Watchdog use the static Enable() method of this class.
+ * This enables and starts all Watchdog object counters and from this point all Watchdog objects must be fed within
+ * their feeding windows.
+ *
+ */
+class Watchdog {
+public:
+	/**
+	 * @brief Callback for feeding purposes
+	 *
+	 * This callback will be run when feedingInterval timeout occurs.
+	 * It must call the Feed method within the specific context.
+	 *
+	 * For example to feed an Watchdog object dedicated for Main thread
+	 * you can use Nrf::PostTask method:
+	 *
+	 *  Nrf::PostTask([watchdog] { watchdog->Feed(); });
+	 */
+	using FeedingCallback = void (*)(Watchdog *watchdog);
+
+	/**
+	 * @brief Create new watchdog object
+	 *
+	 * Each Watchdog object contains callback which is called when the feedingInterval time is elapsed.
+	 * If at least one of the watchdog objects is not fed within the {minWindowTime, maxWindowTime} then the
+	 * reboot occurs.
+	 *
+	 * @param minWindowTime Minimum window Watchdog timeout for Watchdog object in milliseconds.
+	 * @param maxWindowTime Maximum window watchdog timeout for Watchdog object in milliseconds.
+	 * @param feedingInterval if the callback argument is provided then after elapsing time in milliseconds of this
+	 * value the callback will be set.
+	 * @param callback FeedingCallback to be called in the specific place to reset the Watchdog timer
+	 */
+	Watchdog(uint32_t minWindowTime, uint32_t maxWindowTime, uint32_t feedingInterval, FeedingCallback callback);
+
+	/**
+	 * @brief Disable and remove an Watchdog object
```

Comment:
Currently the destructor does nothing.

Reply:
Currently, because I'm still working on that :)